contracts/mechs/nevermined_token/BalanceTrackerNvmSubscriptionToken.sol

```diff
@@ -0,0 +1,161 @@
+        }
+
+        // Calculate how many actual credits need to burn accounting for discount balance
+        if (balance > maxDeliveryRate) {
+            balance -= maxDeliveryRate;
+        } else {
+            maxDeliveryRate -= balance;
+            balance = 0;
+            IERC1155(subscriptionNFT).burn(requester, subscriptionTokenId, maxDeliveryRate);
+
+            emit RequesterCreditsRedeemed(requester, maxDeliveryRate);
+        }
+
+        return balance;
+    }
+
+    /// @dev Gets required token funds.
+    /// @return Received amount.
+    function _getRequiredFunds(address, uint256) internal pure virtual override returns (uint256) {
+        return 0;
+    }
+
+    /// @dev Process mech payment.
+    /// @param mech Mech address.
+    /// @return mechPayment Mech payment.
+    function _processPayment(address mech) internal virtual override returns (uint256, uint256) {
+        // Get mech balance
+        uint256 balance = mapMechBalances[mech];
+        // Check for zero value
+        if (balance == 0) {
+            revert ZeroValue();
+        }
+
+        // Convert mech credits balance into tokens
+        balance = (balance * tokenCreditRatio) / 1e18;
+        mapMechBalances[mech] = balance;
+
+        // Check current contract balance
+        if (balance > trackerBalance) {
+            revert Overflow(balance, trackerBalance);
+        }
+
+        // Proceed with the default mech payment logic
+        return super._processPayment(mech);
+    }
+
+    /// @dev Sets subscription.
+    /// @param _subscriptionNFT Subscription NFT address.
+    /// @param _subscriptionTokenId Subscription token Id.
+    /// @param _tokenCreditRatio Token to credit ratio in 1e18 form.
+    function setSubscription(address _subscriptionNFT, uint256 _subscriptionTokenId, uint256 _tokenCreditRatio) external {
+        // Check for ownership
+        if (msg.sender != owner) {
+            revert OwnerOnly(msg.sender, owner);
+        }
+
+        if (_subscriptionNFT == address(0)) {
+            revert ZeroAddress();
+        }
+
+        if (_subscriptionTokenId == 0 || _tokenCreditRatio == 0) {
+            revert ZeroValue();
+        }
+
+        subscriptionNFT = _subscriptionNFT;
+        subscriptionTokenId = _subscriptionTokenId;
+        tokenCreditRatio = _tokenCreditRatio;
+
+        // Reset owner after setting subscription params
+        owner = address(0);
+
+        emit SubscriptionSet(_subscriptionNFT, _subscriptionTokenId);
+    }
+
+    /// @dev Deposits token funds for requester.
+    /// @param amount Token amount.
+    function deposit(uint256 amount) external pure virtual override {
+        revert NoDepositAllowed(amount);
+    }
+
+    /// @dev Deposits token funds for requester.
+    /// @param amount Token amount.
+    function depositFor(address, uint256 amount) external pure virtual override {
+        revert NoDepositAllowed(amount);
+    }
```

Comment:
We can't accept deposits as all the deposits go via subscription purchase, so we need to block those functions.